stop using &mut self in generated service methods.

adimarco opened the pull request:
This is almost done, I'll wrap it up tomorrow.

Fixes #277 - It turns out that the only place the credentials providers stored state and needed a `&mut self` was if they cached credentials to see if they had expired.  I made that functionality optional by creating an `ExpiringProvider` that wraps other `ProvideAwsCredential` implementations if it's wanted.  It uses a `Mutex` to ensure thread safety.

I updated all the codegen and the S3 hand-written code to stop using `&mut self` everywhere.  Still needs some tests and helper updates.

jimmycuadra commented:
I'd say `AutoRefreshingProvider` instead of `ExpiringProvider`. What it does it automatically refresh expired credentials, not expire them.

I still kinda hoped the use of a mutex could be reserved only for multi-threaded use. Auto-refreshing would be useful even on a single thread. Is it worth trying to make this provider generic across two versions, one for single thread use without a mutex, and one for multi-threaded use with one?

luser commented:
Presumably you could make `AutoRefreshingProvider` generic over some trait, and provide two concrete implementations, like:
```rust
struct BaseAutoRefreshingProvider<T: Something> {}
type AutoRefreshingProvider = BaseAutoRefreshingProvider<RefCell>;
type AutoRefreshingProviderSync = BaseAutoRefreshingProvider<Arc<Mutex>>;
```

The former would be `!Sync` due to the `RefCell`, and the latter would be `Sync`.

adimarco commented:
I renamed the struct to `BaseAutoRefreshingProvider<P>` and provided `Sync` and `!Sync` implementations with `Mutex` and `RefCell` that look like:

```
type AutoRefreshingProviderSync<P> = BaseAutoRefreshingProvider<P, Mutex<AwsCredentials>>;
type AutoRefreshingProvider<P> = BaseAutoRefreshingProvider<P, RefCell<AwsCredentials>>;
```

All in all this seems to work like desired, and allows single-threaded implementations to optionally  refresh credentials automatically without potentially blocking on a mutex.

The 'RefCell' does runtime borrow checking, and therefore does involve some overhead, but at least it doesn't potentially block.  @jimmycuadra - did you different ideas about how to go about implementing the generic provider?  Owning a cached set of credentials that are checked for expiration and possibly refreshed obviously requires some flavor of mutability, and since the goal here is to do it without `&mut self`, that basically leaves us with the interior mutability of the `Cell` types.  (I take some comfort in that this is almost literally the textbook case of when to use `RefCell` - ["e.g. caching forces the implementation to perform mutation; or because you must employ mutation to implement a trait method that was originally defined to take &self"](https://doc.rust-lang.org/std/cell/index.html#implementation-details-of-logically-immutable-methods)

The rest of the credentials code is a mess.  I'm going to clean it up a bit while I'm in here.

jimmycuadra commented:
Nope, that sounds perfect! Great work.

adimarco commented:
I added a couple more type aliases for convenience, making it easy to create the types most people are going to want to use.  Those amount to:

```
pub type DefaultCredentialsProvider = AutoRefreshingProvider<ChainProvider>;
pub type DefaultCredentialsProviderSync = AutoRefreshingProviderSync<ChainProvider>;
```

Other than that, I've documented the above and changed our tests to use the new `DefaultCredentialsProvider` instead of the old `ChainProvider` that no longer refreshes automatically.

jimmycuadra commented:
Awesome work! Those type aliases are a really nice addition and makes me feel even better about this change. It just feels right. :}
